spi/imx6ull-qspi/qspi.c

```diff
@@ -0,0 +1,326 @@
+ * Author: Hubert Badocha
+ *
+ * %LICENSE%
+ */
+
+#include <stdio.h>
+#include <stddef.h>
+#include <sys/interrupt.h>
+#include <sys/mman.h>
+#include <errno.h>
+#include <sys/platform.h>
+#include <sys/threads.h>
+#include <stdbool.h>
+#include <string.h>
+#include <phoenix/arch/imx6ull.h>
+#include "qspi.h"
+#include <board_config.h>
+
+/* Default values for not configured boards */
+#ifndef QSPI_FLASH_A1_SIZE
+#define QSPI_FLASH_A1_SIZE (256 * 1024 * 1024)
+#endif
+#ifndef QSPI_FLASH_A2_SIZE
+#define QSPI_FLASH_A2_SIZE (0)
+#endif
+#ifndef QSPI_FLASH_B1_SIZE
+#define QSPI_FLASH_B1_SIZE (0)
+#endif
+#ifndef QSPI_FLASH_B2_SIZE
+#define QSPI_FLASH_B2_SIZE (0)
+#endif
+
+#define QSPI_MMAP_BASE 0x60000000
+
+#define QSPI_BASE ((uint32_t *)0x21E0000)
+
+#define QSPI_LCKCR_ADDR   (qspi_common.base + 0x304 / sizeof(uint32_t))
+#define QSPI_LCKCR_LOCK   0x01
+#define QSPI_LCKCR_UNLOCK 0x02
+
+#define QSPI_LUT_KEY_ADDR (qspi_common.base + 0x300 / sizeof(uint32_t))
+#define QSPI_LUT_KEY      0x5AF05AF0
+
+#define QSPI_LUTn_ADDR(n)      (qspi_common.base + (0x310 + 4 * (n)) / sizeof(uint32_t))
+#define QSPI_SEQ_START_REGN(n) (4 * (n))
+
+#define QSPI_IPCR_ADDR                    (qspi_common.base + 0x08 / sizeof(uint32_t))
+#define QSPI_IPCR_SET_SEQID(reg, seqid)   ((reg) & ~(0x0F << 24)) | ((seqid & 0x0F) << 24)
+#define QSPI_IPCR_SET_IDATSZ(reg, idatsz) ((reg) & ~(0xFFFF)) | (idatsz)
+
+#define QSPI_FLSHCR_ADDR (qspi_common.base + 0x0C / sizeof(uint32_t))
+
+#define QSPI_MCR_ADDR     (qspi_common.base)
+#define QSPI_MCR_CLR_TXF  (0x01 << 11)
+#define QSPI_MCR_CLR_RXF  (0x01 << 10)
+#define QSPI_MCR_CLR_MDIS (0x01 << 14)
+
+#define QSPI_BFGENCR_ADDR (qspi_common.base + 0x20 / sizeof(uint32_t))
+
+#define QSPI_SR_ADDR   (qspi_common.base + 0x15C / sizeof(uint32_t))
+#define QSPI_SR_BUSY   (0x01)
+#define QSPI_SR_RXWE   (0x01 << 16)
+#define QSPI_SR_TXEDA  (0x01 << 24)
+#define QSPI_SR_TXFULL (0x01 << 27)
+
+#define QSPI_RBDRn_ADRR(n) (qspi_common.base + (0x200 + 4 * (n)) / sizeof(uint32_t))
+
+#define QSPI_RBSR_ADRR    (qspi_common.base + 0x10C / sizeof(uint32_t))
+#define QSPI_RBSR_TO_READ (((*QSPI_RBSR_ADRR) >> 8) & 0x3F)
+
+#define QSPI_SFAR_ADDR (qspi_common.base + 0x100 / sizeof(uint32_t))
+
+#define QSPI_RBCT_ADDR  (qspi_common.base + 0x110 / sizeof(uint32_t))
+#define QSPI_RBCT_RXBRD (0x01 << 8)
+
+#define QSPI_FR_ADDR (qspi_common.base + 0x160 / sizeof(uint32_t))
+#define QSPI_FR_RBDF (0x01 << 16)
+
+#define QSPI_TBDR_ADDR (qspi_common.base + 0x154 / sizeof(uint32_t))
+
+#define QSPI_SFA1AD (qspi_common.base + 0x180 / sizeof(uint32_t))
+#define QSPI_SFA2AD (qspi_common.base + 0x184 / sizeof(uint32_t))
+#define QSPI_SFB1AD (qspi_common.base + 0x188 / sizeof(uint32_t))
+#define QSPI_SFB2AD (qspi_common.base + 0x18C / sizeof(uint32_t))
+
+#define WATERMARK 16
+
+#define NUM_LUT_SEQ 16
+
+typedef struct {
+	int mux;
+	char val;
+	char sion;
+} qspi_pctl_mux_t;
+
+typedef struct {
+	int pad;
+	char speed;
+	char dse;
+	char sre;
+} qspi_pctl_pad_t;
```

Comment:
these structs are used only exactly once - no need to define new types for them, just create anonymous structs:
```suggestion
static const struct {
	int mux;
	char val;
	char sion;
} qspi_pctl_mux = {
...
};

static const struct {
	int pad;
	char speed;
	char dse;
	char sre;
} qspi_pctl_pad = { ....
```